src/client/dfuse/pil4dfs/hook.c

```diff
@@ -172,77 +179,88 @@ read_map_file(char **buf, const char str_target[], char **pos)
+	path_ld[end - start] = 0;
+
+	pos = strstr(read_buff_map, "/lib64/libc.so");
+	if (pos == NULL)
+		/* try a different format */
+		pos = strstr(read_buff_map, "/lib64/libc-2");
+	if (pos == NULL) {
+		printf("Warning: Failed to find libc.so.\n");
+		goto err;
 	}
-	if (pStart == NULL) {
-		printf("Fail to determine the starting position of libc path.\nQuit\n");
-		exit(1);
+	get_string_pos(pos, &start, &end);
+	if ((end - start + 1) >= PATH_MAX) {
+		printf("The length of path_libc is too long!\nQuit\n");
+		goto err;
 	}
-
-	if ((long int)(pEnd - pStart) >= PATH_MAX) {
-		printf("The length of path_ld is too long!\nQuit\n");
-		exit(1);
+	path_libc = malloc(end - start + 1);
+	if (path_libc == NULL) {
+		printf("Warning: Failed to allocate memory for path_libc.\n");
+		goto err;
 	}
-	path_ld = malloc((int)(pEnd - pStart) + 1);
-	if (path_ld == NULL) {
-		printf("Failed to allocate memory for path_ld.\nQuit\n");
-		exit(1);
+	_Pragma("GCC diagnostic push")
+	_Pragma("GCC diagnostic ignored \"-Warray-bounds\"")
+	_Pragma("GCC diagnostic ignored \"-Wstringop-overflow\"")
+	_Pragma("GCC diagnostic ignored \"-Wrestrict\"")
+	memcpy(path_libc, start, end - start);
+	path_libc[end - start] = 0;
+	_Pragma("GCC diagnostic pop")
+
+	/* compose path_libpthread with path_libc string */
+	if ((end - start + 1 + 6) >= PATH_MAX) {
+		printf("The length of path_libpthread is too long!\nQuit\n");
+		goto err;
+	}
+	path_libpthread = malloc(end - start + 1 + 6);
+	if (path_libpthread == NULL) {
+		printf("Warning: Failed to allocate memory for path_libpthread.\n");
+		goto err;
 	}
-	memcpy(path_ld, pStart, pEnd - pStart);
-	path_ld[pEnd - pStart] = 0;
 
-	memcpy(lib_ver_str, pPos + 4, pEnd - 3 - (pPos + 4));
-	lib_ver_str[pEnd - 3 - (pPos + 4)] = 0;
-	memcpy(lib_dir_str, pStart, pPos - pStart);
-	lib_dir_str[pPos - pStart] = 0;
+	memcpy(path_libpthread, start, end - start);
+	path_libpthread[end - start + 1 + 5] = 0;
+	pos = strstr(path_libpthread, "libc");
+	memcpy(pos + 3    , "pthread", 7);
+	memcpy(pos + 3 + 7, path_libc + (pos - path_libpthread + 4), end - start + 1 + 6 -
+	       (pos - path_libpthread + 4));
+	if (path_libpthread[end - start + 1 + 4] >= '5' &&
+	    path_libpthread[end - start + 1 + 4] <= '7') {
+		path_libpthread[end - start + 1 + 4] = '0';
+	}
 
 	free(read_buff_map);
+	found_libc = 1;
+	return;
 
-	rc = asprintf(&path_libc, "%s/libc-%s.so", lib_dir_str, lib_ver_str);
-	if (rc < 0) {
-		printf("Failed to allocate memory for path_libc!\nQuit\n");
-		exit(1);
-	}
-	if (rc >= PATH_MAX) {
-		printf("The length of path_libc is too long!\nQuit\n");
-		exit(1);
-	}
-
-	rc = asprintf(&path_libpthread, "%s/libpthread-%s.so", lib_dir_str, lib_ver_str);
-	if (rc < 0) {
-		printf("Failed to allocate memory for path_libpthread!\nQuit\n");
-		exit(1);
-	}
-	if (rc >= PATH_MAX) {
-		printf("The length of path_libpthread is too long!\nQuit\n");
-		exit(1);
-	}
+err:
+	free(read_buff_map);
+	found_libc = 0;
+	exit(1);
```

Comment:
There are lots of memory leaks here, and setting found_libc seems redundant.

This seems like it would be a great place to perform some common 'cannot continue` logging that we could check for in the testing, then call `daos_debug_fini` before exiting.

Reply:
libpil4dfs's destructor function will still be called inside exit(). free_memory_in_hook() will be called when install_hook() fails.